src/devices/sound/stt_sa1.cpp

```diff
@@ -0,0 +1,210 @@
+				m_voice[v].addr_start = (m_voice[v].addr_start & 0x0000ffff000) | (uint64_t(data) << 28);
+				LOG("voice %d: addr_start = %08llx\n", v, m_voice[v].addr_start >> 12);
+				break;
+
+			case 6:
+				m_voice[v].addr_end = (m_voice[v].addr_end & 0xffff0000000) | (uint64_t(data) << 12);
+				LOG("voice %d: addr_end = %08llx\n", v, m_voice[v].addr_end >> 12);
+				break;
+			case 7:
+				m_voice[v].addr_end = (m_voice[v].addr_end & 0x0000ffff000) | (uint64_t(data) << 28);
+				LOG("voice %d: addr_end = %08llx\n", v, m_voice[v].addr_end >> 12);
+				break;
+
+			case 0x0b:
+				m_voice[v].vol_l = data;
+				LOG("voice %d: vol_l = %08x\n", v, m_voice[v].vol_l);
+				break;
+			case 0x0c:
+				m_voice[v].vol_r = data;
+				LOG("voice %d: vol_r = %08x\n", v, m_voice[v].vol_r);
+				break;
+
+			default:
+				LOG("Unknown register usage: voice %d, register %x, data %04x\n", v, reg, data);
+				break;
+		}
+	}
+	else
+	{
+		put_u16be(&m_ram[offset * 2], data);
+	}
+}
+
+uint16_t stt_sa1_device::key_r()
+{
+	if (!m_enabled)
+		return 0;
+
+	m_stream->update();
+	return m_keyctrl;
+}
+
+void stt_sa1_device::key_w(uint16_t data)
+{
+	if (!m_enabled)
+		return;
+
+	const u16 prev = m_keyctrl;
+
+	m_stream->update();
+
+	m_keyctrl = data;
+
+	for (int v = 0; v < 8; v++) {
+		if (BIT(m_keyctrl, v) && !BIT(prev, v)) {
+			// keyon
+			m_voice[v].enabled = true;
+			m_voice[v].addr_cur = m_voice[v].addr_start;
+		} else if (!BIT(m_keyctrl, v) && BIT(prev, v)) {
+			// keyoff
+			m_voice[v].enabled = false;
+		}
+	}
+}
+
+int8_t stt_sa1_device::read_sample(offs_t offset)
+{
+	if (offset >= 0x600000)
+		return m_readsample_cb(offset); // Held externally in flash chips
+
+	return m_ram[offset];
+}
+
+void stt_sa1_device::device_start()
+{
+	m_stream = stream_alloc(0, 2, clock() / 448);
+	m_ram = make_unique_clear<uint8_t[]>(0x600000); // 12 512k x 8 SRAM chips
+
+	save_item(STRUCT_MEMBER(m_voice, addr_start));
+	save_item(STRUCT_MEMBER(m_voice, addr_cur));
+	save_item(STRUCT_MEMBER(m_voice, addr_end));
+	save_item(STRUCT_MEMBER(m_voice, vol_l));
+	save_item(STRUCT_MEMBER(m_voice, vol_r));
+	save_item(STRUCT_MEMBER(m_voice, freq));
+	save_item(STRUCT_MEMBER(m_voice, is_looped));
+	save_item(STRUCT_MEMBER(m_voice, enabled));
+	save_item(NAME(m_regs));
+	save_item(NAME(m_enabled));
+	save_pointer(NAME(m_ram), 0x600000);
+}
+
+void stt_sa1_device::device_reset()
+{
+	m_enabled = false;
+
+	std::fill(std::begin(m_regs), std::end(m_regs), 0);
+
+	for (int i = 0; i < 8; i++)
+	{
+		m_voice[i].addr_start = 0;
+		m_voice[i].addr_cur = 0;
+		m_voice[i].addr_end = 0;
+		m_voice[i].vol_l = 0;
+		m_voice[i].vol_r = 0;
+		m_voice[i].freq = 0;
+		m_voice[i].is_looped = false;
+		m_voice[i].enabled = false;
+	}
+}
+
+void stt_sa1_device::sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs)
+{
+	outputs[0].fill(0);
+	outputs[1].fill(0);
+
+	for (int v = 0; v < 8; v++)
+	{
+		voice_t *voice = &m_voice[v];
```

Comment:
You can use a reference for `voice` (like `voice_t &voice = m_voice[v];`) since you shouldn’t need to index or modify the pointer.